include/RevFenv.h

```diff
@@ -11,69 +11,103 @@
 #ifndef _REV_FENV_H_
 #define _REV_FENV_H_
 
+#include "RevInstTable.h"
 #include <cfenv>
-#include <memory>
+#include <cmath>
+#include <stdexcept>
 
 // GCC and Clang do not fully support FENV_ACCESS now. See:
 //
+// https://bugs.llvm.org/show_bug.cgi?id=8100
 // https://gcc.gnu.org/legacy-ml/gcc-patches/2003-09/msg00104.html
 // https://gcc.gcc.gnu.narkive.com/hDi1eOot/setting-frounding-math-by-default
-// https://bugs.llvm.org/show_bug.cgi?id=8100
-// https://github.com/llvm/llvm-project/issues/8472
 // https://discourse.llvm.org/t/why-is-pragma-stdc-fenv-access-not-supported/46128/25
 //
 // Currently FP_MODE_FLAG in CMakeLists.txt is used as a workaround
 //#pragma STDC FENV_ACCESS ON
 
 namespace SST::RevCPU {
 
-// TODO: Right now we only need to save/restore rounding mode.
-// Later, we may need to save and restore the entire fenv_t state
 class RevFenv {
-  int saved_round;  ///< Saved Floating-Point Rounding Mode
-
-  // We use Meyers singleton to avoid initialization order fiasco
-  static int& round() {
-    thread_local int round = fegetround();
-    return round;
-  }
+  FCSR&       fcsr;         // Reference to this register file's FCSR
+  std::fenv_t saved_env{};  // The saved FP environment which is restored
 
 public:
-  /// Constructor saves Fenv state
-  RevFenv() : saved_round( round() ) {
-    if( saved_round < 0 ) {
-      throw std::runtime_error( "Getting floating-point rounding mode with "
-                                "fegetround() is not working." );
+  /// Constructor saves Fenv state to be restored at destruction
+  RevFenv( RevRegFile* R, FRMode rm, SST::Output* output ) : fcsr( R->GetFCSR() ) {
+
+    // Save host's FP environment and set flags to default, non-trapping
+    if( std::feholdexcept( &saved_env ) ) {
+      throw std::runtime_error( "Getting floating-point environment with feholdexcept() is not working." );
+    }
+
+    // If the encoded rounding mode is dynamic, load the frm register
+    if( rm == FRMode::DYN ) {
+      rm = R->GetFRM();
+    }
+
+    // Set the Floating-Point Rounding Mode on the host
+    int ret = 0;
+    switch( rm ) {
+    case FRMode::RNE:  // Round to Nearest, ties to Even
+      ret = std::fesetround( FE_TONEAREST );
+      break;
+
+    case FRMode::RTZ:  // Round towards Zero
+      ret = std::fesetround( FE_TOWARDZERO );
+      break;
+
+    case FRMode::RDN:  // Round Down (towards -Inf)
+      ret = std::fesetround( FE_DOWNWARD );
+      break;
+
+    case FRMode::RUP:  // Round Up (towards +Inf)
+      ret = std::fesetround( FE_UPWARD );
+      break;
+
+    case FRMode::RMM:  // Round to Nearest, ties to Max Magnitude
+#ifdef FE_TONEARESTFROMZERO
+      ret = std::fesetround( FE_TONEARESTFROMZERO );
+#else
+      output->fatal( CALL_INFO, -1, "Error: Round to nearest Max Magnitude not implemented at PC = 0x%" PRIx64 "\n", R->GetPC() );
+#endif
```

Comment:
Round to nearest Max Magnitude, also called roundTiesToAway, is optional on binary radix floating-point systems, and is only mandatory on decimal radix floating-point systems, because of its use in financial calculations.

A C2X proposal added the `FE_TONEARESTFROMZERO` as an optional macro defining the rounding mode. If this macro is supported on the host environment, we can support RISC-V Round to Nearest Max Magnitude; otherwise it's not implemented.